content/courses/advanced-ada/parts/resource_management/controlled_types.rst

```diff
@@ -1189,13 +1189,311 @@ This diagram shows the initialization sequence:
+       begin
+          Put_Line ("Adjust: Sub_1...");
+       end Adjust;
+
+       procedure Finalize (E : in out Sub_1) is
+       begin
+          Put_Line ("Finalize: Sub_1...");
+       end Finalize;
+
+       procedure Initialize (E : in out Sub_2) is
+       begin
+          Put_Line ("Initialize: Sub_2...");
+       end Initialize;
+
+       procedure Adjust (E : in out Sub_2) is
+       begin
+          Put_Line ("Adjust: Sub_2...");
+       end Adjust;
+
+       procedure Finalize (E : in out Sub_2) is
+       begin
+          Put_Line ("Finalize: Sub_2...");
+       end Finalize;
+
+    end Subs;
+
+Next, we override the :ada:`Adjust` procedure of the :ada:`T` type from the
+:ada:`Simple_Controlled_Types` package:
+
+.. code:: ada compile_button project=Courses.Advanced_Ada.Resource_Management.Controlled_Types.Adjustment.Controlled_Initialization
+
+    with Ada.Finalization;
+
+    with Subs; use Subs;
+
+    package Simple_Controlled_Types is
+
+       type T is tagged private;
+
+       procedure Dummy (E : T);
+
+    private
+
+       function Default_Init return Integer;
+
+       type T is new
+         Ada.Finalization.Controlled with
+       record
+          S1 : Sub_1;
+          S2 : Sub_2;
+          I  : Integer := Default_Init;
+       end record;
+
+       overriding
+       procedure Initialize (E : in out T);
+
+       overriding
+       procedure Adjust (E : in out T);
+
+       overriding
+       procedure Finalize (E : in out T);
+
+    end Simple_Controlled_Types;
+
+    with Ada.Text_IO; use Ada.Text_IO;
+
+    package body Simple_Controlled_Types is
+
+       function Default_Init return Integer is
+       begin
+          Put_Line ("Default_Init: Integer...");
+          return 42;
+       end Default_Init;
+
+       procedure Dummy (E : T) is
+       begin
+          Put_Line ("(Dummy: T...)");
+       end Dummy;
+
+       procedure Initialize (E : in out T) is
+       begin
+          Put_Line ("Initialize: T...");
+       end Initialize;
+
+       procedure Adjust (E : in out T) is
+       begin
+          Put_Line ("Adjust: T...");
+       end Adjust;
+
+       procedure Finalize (E : in out T) is
+       begin
+          Put_Line ("Finalize: T...");
+       end Finalize;
+
+    end Simple_Controlled_Types;
+
+Finally, this is the main application:
+
+.. code:: ada run_button project=Courses.Advanced_Ada.Resource_Management.Controlled_Types.Adjustment.Controlled_Initialization
+
+    with Ada.Text_IO; use Ada.Text_IO;
+
+    with Simple_Controlled_Types;
+    use  Simple_Controlled_Types;
+
+    procedure Show_Controlled_Types is
+       A, B : T;
+    begin
+       Dummy (A);
+
+       Put_Line ("----------");
+       Put_Line ("A := B");
+       A := B;
+       Put_Line ("----------");
+    end Show_Controlled_Types;
+
+When running this code, we see that the :ada:`S1` and :ada:`S2` components are
+adjusted before the adjustment of the parent type :ada:`T` takes place.
+
```

Comment:
Consider saying something to the effect that the components must be done before the enclosing object because the semantics of the adjustment of the whole might depend on the states of the parts (the components) so those states must already be in place.

Reply:
Yes, good point! I've just added this phrase (up in the introduction of this subsection).